TypeScript errors in 12.0.0

After upgrading to Shepherd.js 12.0.0, TypeScript no longer compiles my code that uses `shepherd.js`. It fails with the following error:

> error TS2307: Cannot find module 'shepherd.js' or its corresponding type declarations.

For reference, [Are the Types Wrong](https://github.com/arethetypeswrong/arethetypeswrong.github.io) is reporting additional problems for 12.0.0, compared to 11.2.0:

* https://arethetypeswrong.github.io/?p=shepherd.js%4012.0.0
* https://arethetypeswrong.github.io/?p=shepherd.js%4011.2.0

@RobbieTheWagner and @chuckcarpenter , I'm getting a different set of errors now:

```
error TS2307: Cannot find module 'shepherd.js' or its corresponding type declarations.
  There are types at '/builds/group/project/node_modules/shepherd.js/dist/shepherd.d.ts', but this result could not be resolved under your current 'moduleResolution' setting. Consider updating to 'node16', 'nodenext', or 'bundler'.
```

My tsconfig.json `moduleResultion` is `node`, but  [Are the Types Wrong](https://arethetypeswrong.github.io/?p=shepherd.js%4012.0.1) is reporting several errors, which makes me think that changing my `moduleResolution` wouldn't necessarily help:

  | "shepherd.js" | "shepherd.js/*"
-- | -- | --
node10 | 💀 Resolution failed | (wildcard)
node16 (from CJS) | ❗️ Incorrect default export🥴 Internal resolution error | (wildcard)
node16 (from ESM) | 🥴 Internal resolution error🎭 Masquerading as CJS | (wildcard)
bundler | 🥴 Internal resolution error | (wildcard)

--- a/landing/src/layouts/MainPage.astro
+++ b/landing/src/layouts/MainPage.astro
@@ -198,6 +198,10 @@ const shepherdIncludeCode = `
     }
 
     function setupShepherd() {
+      const element = document.createElement('p');
+      element.innerText =
+        'Including Shepherd is easy! Just include shepherd.js. The styles are bundled with the JS.';
+
       const shepherd = new Shepherd.Tour({
         id: 'ts4udfrnm7o95ny7e9qbjo3t',
         defaultStepOptions: {
@@ -312,10 +316,6 @@ const shepherdIncludeCode = `
         ],
         useModalOverlay: true
       });
-
-      const element = document.createElement('p');
-      element.innerText =
-        'Including Shepherd is easy! Just include shepherd.js. The styles are bundled with the JS.';
 
       // This should add steps after the ones added with `addSteps`
       shepherd.addStep({

--- a/landing/src/layouts/MainPage.astro
+++ b/landing/src/layouts/MainPage.astro
@@ -182,8 +182,6 @@ const shepherdIncludeCode = `
   'use strict';
 
   import Shepherd from 'shepherd.js';
-  import type { StepOptions } from 'shepherd.js/dist/step';
-  import type { Tour } from 'shepherd.js/dist/tour';
 
   (function () {
     async function init() {
@@ -223,19 +221,93 @@ const shepherdIncludeCode = `
             buttons: [
               {
                 action() {
-                  return (this as unknown as Tour).cancel();
+                  return this.cancel();
                 },
                 secondary: true,
                 text: 'Exit'
               },
               {
                 action() {
-                  return (this as unknown as Tour).next();
+                  return this.next();
                 },
                 text: 'Next'
               }
             ],
             id: 'welcome'
+          },
+          {
+            title: 'Including',
+            text: element,
+            attachTo: {
+              element: '.hero-including',
+              on: 'bottom'
+            },
+            buttons: [
+              {
+                action() {
+                  return this.back();
+                },
+                secondary: true,
+                text: 'Back'
+              },
+              {
+                action() {
+                  return this.next();
+                },
+                text: 'Next'
+              }
+            ],
+            id: 'including'
+          },
+          {
+            title: 'Creating a Shepherd Tour',
+            text:
+              'Creating a Shepherd tour is easy. too! ' +
+              'Just create a `Tour` instance, and add as many steps as you want.',
+            attachTo: {
+              element: '.hero-example',
+              on: 'right'
+            },
+            buttons: [
+              {
+                action() {
+                  return this.back();
+                },
+                secondary: true,
+                text: 'Back'
+              },
+              {
+                action() {
+                  return this.next();
+                },
+                text: 'Next'
+              }
+            ],
+            id: 'creating'
+          },
+          {
+            title: 'Attaching to Elements',
+            text: 'Your tour steps can target and attach to elements in DOM (like this step).',
+            attachTo: {
+              element: '.hero-example',
+              on: 'left'
+            },
+            buttons: [
+              {
+                action() {
+                  return this.back();
+                },
+                secondary: true,
+                text: 'Back'
+              },
+              {
+                action() {
+                  return this.next();
+                },
+                text: 'Next'
+              }
+            ],
+            id: 'attaching'
           }
         ],
         useModalOverlay: true
@@ -245,86 +317,6 @@ const shepherdIncludeCode = `
       element.innerText =
         'Including Shepherd is easy! Just include shepherd.js. The styles are bundled with the JS.';
 
-      // These steps should be added via `addSteps`
-      const steps = [
-        {
-          title: 'Including',
-          text: element,
-          attachTo: {
-            element: '.hero-including',
-            on: 'bottom'
-          },
-          buttons: [
-            {
-              action() {
-                return (this as unknown as Tour).back();
-              },
-              secondary: true,
-              text: 'Back'
-            },
-            {
-              action() {
-                return (this as unknown as Tour).next();
-              },
-              text: 'Next'
-            }
-          ],
-          id: 'including'
-        },
-        {
-          title: 'Creating a Shepherd Tour',
-          text:
-            'Creating a Shepherd tour is easy. too! ' +
-            'Just create a `Tour` instance, and add as many steps as you want.',
-          attachTo: {
-            element: '.hero-example',
-            on: 'right'
-          },
-          buttons: [
-            {
-              action() {
-                return (this as unknown as Tour).back();
-              },
-              secondary: true,
-              text: 'Back'
-            },
-            {
-              action() {
-                return (this as unknown as Tour).next();
-              },
-              text: 'Next'
-            }
-          ],
-          id: 'creating'
-        },
-        {
-          title: 'Attaching to Elements',
-          text: 'Your tour steps can target and attach to elements in DOM (like this step).',
-          attachTo: {
-            element: '.hero-example',
-            on: 'left'
-          },
-          buttons: [
-            {
-              action() {
-                return (this as unknown as Tour).back();
-              },
-              secondary: true,
-              text: 'Back'
-            },
-            {
-              action() {
-                return (this as unknown as Tour).next();
-              },
-              text: 'Next'
-            }
-          ],
-          id: 'attaching'
-        }
-      ] as StepOptions[];
-      // @ts-ignore
-      shepherd.addSteps(steps);
-
       // This should add steps after the ones added with `addSteps`
       shepherd.addStep({
         title: 'Centered Shepherd Element',
@@ -332,14 +324,14 @@ const shepherdIncludeCode = `
         buttons: [
           {
             action() {
-              return (this as unknown as Tour).back();
+              return this.back();
             },
             secondary: true,
             text: 'Back'
           },
           {
             action() {
-              return (this as unknown as Tour).next();
+              return this.next();
             },
             text: 'Next'
           }
@@ -357,14 +349,14 @@ const shepherdIncludeCode = `
         buttons: [
           {
             action() {
-              return (this as unknown as Tour).back();
+              return this.back();
             },
             secondary: true,
             text: 'Back'
           },
           {
             action() {
-              return (this as unknown as Tour).next();
+              return this.next();
             },
             text: 'Done'
           }

--- a/landing/src/layouts/MainPage.astro
+++ b/landing/src/layouts/MainPage.astro
@@ -182,8 +182,8 @@ const shepherdIncludeCode = `
   'use strict';
 
   import Shepherd from 'shepherd.js';
-  import type { StepOptions } from 'shepherd.js/step';
-  import type { Tour } from 'shepherd.js/tour';
+  import type { StepOptions } from 'shepherd.js/dist/step';
+  import type { Tour } from 'shepherd.js/dist/tour';
 
   (function () {
     async function init() {
@@ -246,7 +246,7 @@ const shepherdIncludeCode = `
         'Including Shepherd is easy! Just include shepherd.js. The styles are bundled with the JS.';
 
       // These steps should be added via `addSteps`
-      const steps: Array<StepOptions> = [
+      const steps = [
         {
           title: 'Including',
           text: element,
@@ -321,8 +321,8 @@ const shepherdIncludeCode = `
           ],
           id: 'attaching'
         }
-      ];
-
+      ] as StepOptions[];
+      // @ts-ignore
       shepherd.addSteps(steps);
 
       // This should add steps after the ones added with `addSteps`

--- a/shepherd.js/.eslintignore
+++ b/shepherd.js/.eslintignore
@@ -1,1 +1,2 @@
 /dist/
+/scripts/

--- a/shepherd.js/package.json
+++ b/shepherd.js/package.json
@@ -55,8 +55,7 @@
   "dependencies": {
     "@floating-ui/dom": "^1.5.3",
     "deepmerge-ts": "^5.1.0",
-    "idb": "^8.0.0",
-    "rollup-plugin-dts": "^6.1.0"
+    "idb": "^8.0.0"
   },
   "devDependencies": {
     "@babel/core": "^7.24.5",
@@ -79,6 +78,7 @@
     "rimraf": "^5.0.5",
     "rollup": "^4.17.2",
     "rollup-plugin-analyzer": "^4.0.0",
+    "rollup-plugin-copy": "^3.5.0",
     "rollup-plugin-filesize": "^10.0.0",
     "rollup-plugin-license": "^3.3.1",
     "rollup-plugin-livereload": "^2.0.5",

--- a/shepherd.js/package.json
+++ b/shepherd.js/package.json
@@ -18,20 +18,19 @@
     "Robbie Wagner <rwwagner90@gmail.com>",
     "Chuck Carpenter <chuck@shipshape.io>"
   ],
+  "main": "dist/shepherd.js",
+  "module": "dist/shepherd.mjs",
   "exports": {
     ".": {
-      "types": "./dist/shepherd.d.ts",
-      "require": "./dist/shepherd.js",
       "import": "./dist/shepherd.mjs",
-      "default": "./dist/shepherd.js"
+      "require": "./dist/shepherd.js"
     },
     "./*": {
-      "types": "./dist/*.d.ts",
-      "require": "./dist/*.js",
       "import": "./dist/*.mjs",
-      "default": "./dist/*.js"
+      "require": "./dist/*.js"
     }
   },
+  "types": "shepherd.d.ts",
   "typesVersions": {
     "*": {
       "*": [
@@ -40,7 +39,7 @@
     }
   },
   "scripts": {
-    "build": "pnpm clean && rollup -c",
+    "build": "pnpm clean && rollup -c && node scripts/postprocess.js",
     "clean": "rimraf ./dist",
     "esdoc": "esdoc",
     "lint": "concurrently \"npm:lint:*(!fix)\" --names \"lint:\"",
@@ -56,7 +55,8 @@
   "dependencies": {
     "@floating-ui/dom": "^1.5.3",
     "deepmerge-ts": "^5.1.0",
-    "idb": "^8.0.0"
+    "idb": "^8.0.0",
+    "rollup-plugin-dts": "^6.1.0"
   },
   "devDependencies": {
     "@babel/core": "^7.24.5",

--- a/shepherd.js/rollup.config.mjs
+++ b/shepherd.js/rollup.config.mjs
@@ -2,7 +2,7 @@ import autoprefixer from 'autoprefixer';
 import fs from 'fs';
 import cssnanoPlugin from 'cssnano';
 import { babel } from '@rollup/plugin-babel';
-import dts from 'rollup-plugin-dts';
+import copy from 'rollup-plugin-copy';
 import serve from 'rollup-plugin-serve';
 import livereload from 'rollup-plugin-livereload';
 import filesize from 'rollup-plugin-filesize';
@@ -46,7 +46,18 @@ const plugins = [
     banner
   }),
   filesize(),
-  visualizer()
+  visualizer(),
+  copy({
+    targets: [
+      {
+        src: 'dist/shepherd.d.ts',
+        dest: 'dist',
+        rename: 'shepherd.d.mts'
+      }
+    ],
+    copySync: true,
+    hook: 'writeBundle' // Ensure copying happens after the build is complete
+  })
 ];
 
 // If we are running with --environment DEVELOPMENT, serve via browsersync for local development
@@ -77,30 +88,5 @@ export default [
       }
     ],
     plugins
-  },
-  {
-    input: 'src/shepherd.ts',
-    output: { file: 'dist/shepherd.d.mts', format: 'es' },
-    plugins: [
-      dts({
-        compilerOptions: {
-          emitDeclarationOnly: true
-        }
-      })
-    ]
-  },
-  {
-    input: 'src/shepherd.ts',
-    output: { file: 'dist/shepherd.d.ts', format: 'cjs' },
-    plugins: [
-      dts({
-        compilerOptions: {
-          emitDeclarationOnly: true,
-          module: 'CommonJS',
-          esModuleInterop: false,
-          verbatimModuleSyntax: false,
-        }
-      })
-    ]
   }
 ];

--- a/shepherd.js/rollup.config.mjs
+++ b/shepherd.js/rollup.config.mjs
@@ -2,6 +2,7 @@ import autoprefixer from 'autoprefixer';
 import fs from 'fs';
 import cssnanoPlugin from 'cssnano';
 import { babel } from '@rollup/plugin-babel';
+import dts from 'rollup-plugin-dts';
 import serve from 'rollup-plugin-serve';
 import livereload from 'rollup-plugin-livereload';
 import filesize from 'rollup-plugin-filesize';
@@ -71,9 +72,35 @@ export default [
         dir: 'dist',
         entryFileNames: '[name].js',
         format: 'cjs',
-        sourcemap: true
+        sourcemap: true,
+        exports: 'auto'
       }
     ],
     plugins
+  },
+  {
+    input: 'src/shepherd.ts',
+    output: { file: 'dist/shepherd.d.mts', format: 'es' },
+    plugins: [
+      dts({
+        compilerOptions: {
+          emitDeclarationOnly: true
+        }
+      })
+    ]
+  },
+  {
+    input: 'src/shepherd.ts',
+    output: { file: 'dist/shepherd.d.ts', format: 'cjs' },
+    plugins: [
+      dts({
+        compilerOptions: {
+          emitDeclarationOnly: true,
+          module: 'CommonJS',
+          esModuleInterop: false,
+          verbatimModuleSyntax: false,
+        }
+      })
+    ]
   }
 ];

--- a/shepherd.js/scripts/postprocess.js
+++ b/shepherd.js/scripts/postprocess.js
@@ -3,8 +3,5 @@ const path = require('path');
 
 const declarationFile = path.join(__dirname, '../dist', 'shepherd.d.ts');
 let content = fs.readFileSync(declarationFile, 'utf8');
-content = content.replace(
-  /export { Shepherd as default }/g,
-  'export = Shepherd'
-);
+content = content.replace(/export default Shepherd/g, 'export = Shepherd');
 fs.writeFileSync(declarationFile, content);

--- a/shepherd.js/scripts/postprocess.js
+++ b/shepherd.js/scripts/postprocess.js
@@ -1,0 +1,10 @@
+const fs = require('fs');
+const path = require('path');
+
+const declarationFile = path.join(__dirname, '../dist', 'shepherd.d.ts');
+let content = fs.readFileSync(declarationFile, 'utf8');
+content = content.replace(
+  /export { Shepherd as default }/g,
+  'export = Shepherd'
+);
+fs.writeFileSync(declarationFile, content);

--- a/shepherd.js/src/step.ts
+++ b/shepherd.js/src/step.ts
@@ -355,6 +355,7 @@ export class Step extends Evented {
    * Hide the step
    */
   hide() {
+    // @ts-expect-error TODO: investigate once Svelte is typed to use Modal component
     this.tour.modal?.hide();
 
     this.trigger('before-hide');
@@ -580,7 +581,7 @@ export class Step extends Evented {
     if (!this.tour.modal) {
       this.tour.setupModal();
     }
-
+    // @ts-expect-error TODO: investigate once Svelte is typed to use Modal component
     this.tour.modal?.setupForStep(this);
     this._styleTargetElementForStep(this);
 

--- a/shepherd.js/src/tour.ts
+++ b/shepherd.js/src/tour.ts
@@ -193,7 +193,7 @@ export class Tour extends Evented {
   currentStep?: Step | null;
   focusedElBeforeOpen?: HTMLElement | null;
   id?: string;
-  modal?: ShepherdModal;
+  modal?: Record<string, unknown> | null;
   options: TourOptions;
   steps: Array<Step>;
 
@@ -502,6 +502,7 @@ export class Tour extends Evented {
     this.trigger('inactive', { tour: this });
 
     if (this.modal) {
+      // @ts-expect-error TODO: investigate once Svelte is typed
       this.modal.hide();
     }
 

--- a/shepherd.js/tsconfig.json
+++ b/shepherd.js/tsconfig.json
@@ -1,9 +1,9 @@
 {
   "$schema": "https://json.schemastore.org/tsconfig",
   "compilerOptions": {
-    "target": "es2021",
-    "module": "esnext",
-    "moduleResolution": "bundler",
+    "target": "es2020",
+    "module": "es2020",
+    "moduleResolution": "node",
     // We don't want to include types dependencies in our compiled output, so tell TypeScript
     // to enforce using `import type` instead of `import` for Types.
     "verbatimModuleSyntax": true,
@@ -43,6 +43,6 @@
       "src/*": ["./src/*"]
     }
   },
-  "include": ["src/**/*"],
+  "include": ["src/**/*", "src/**/*.svelte"],
   "exclude": ["node_modules", "dist"]
 }

